certifikit-cli/src/main/kotlin/app/cash/certifikit/cli/errors/errorCategorisation.kt

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package app.cash.certifikit.cli.errors
+
+import app.cash.certifikit.cli.Main
+import java.io.IOException
+import java.security.cert.CertificateExpiredException
+import javax.net.ssl.SSLHandshakeException
+
+fun Exception.findMatchingCause(matcher: (Throwable) -> Boolean): Throwable? =
+  generateSequence<Throwable>(this) { it.cause }.find { matcher(it) }
+
+fun Main.classify(
+  e: IOException
+): IOException {
+  return when {
+    // java.security.cert.CertificateExpiredException: NotAfter: Mon Apr 13 00:59:59 BST 2015
+    e.findMatchingCause { it is CertificateExpiredException } != null -> CertificationException(
+        "Certificate for $host is expired", e
+    )
+    // javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
+    e.findMatchingCause { it.javaClass.simpleName == "SunCertPathBuilderException" } != null -> CertificationException(
+        "Certificate for $host is untrusted", e
```

Comment:
If we’re gonna go through all this work to interpret certificate failures, we might as well go all the way and have an enum or type hierarchy for the different failure reasons?

(for follow-up, and probably not in the CLI)